### PR TITLE
SOT-127: FRI implementation

### PR DESCRIPTION
**Pull request by VanhGer**

### CODE REVIEW

File: fri/fri.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    cargo run --example fri-example   
+    ```
+The above code will run the `main` function in `example.rs` files located in the `examples`
+directory, which is the example usage of this library.
+
+### How it works
+
+We recommend reading our documentation of FRI 
+[here](https://github.com/sota-zk-labs/zkp-documents/blob/main/docs/fri.md) and slide 
+[here](https://github.com/sota-zk-labs/zkp-documents/blob/main/presentations/fri_implementation.pptx).
```
Comment (ktokto313) [inline fri/fri.md:39]: Invalid link (doesn't work)

File: fri/examples/example.rs [hunk 2]
```diff
@@ -19,12 +20,10 @@ fn main() {
     let blowup_factor: usize = 2;
     let number_of_queries: usize = 2;
     println!("Generate proof...");
-    let proof = generate_proof(&poly, blowup_factor, number_of_queries);
```
Comment (zk-steve) [inline fri/examples/example.rs]: generate_proof should take the poly's reference, not its ownership

File: fri/src/fiat_shamir/transcript.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
+use ark_ff::PrimeField;
```
Comment (zk-steve) [inline fri/src/fiat_shamir/transcript.rs:1]: rearrange the import

File: fri/src/fiat_shamir/transcript.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        self.generate_challenges(number)
+            .into_iter()
+            .map(|field| field.into_bigint().as_ref()[0] as usize)
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Transcript;
```
Comment (zk-steve) [inline fri/src/fiat_shamir/transcript.rs:144]: rearrange the import

File: fri/src/prover.rs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+#[derive(Clone, Debug)]
+pub struct Proof<F: PrimeField> {
+    pub domain_size: usize,
+    pub coset: F,
+    pub number_of_queries: usize,
+    pub layers_root: Vec<F>,
+    pub const_val: F,
+    pub decommitment_list: Vec<Decommitment<F>>,
+    // pub challenge_list: Vec<usize>
```
Comment (zk-steve) [inline fri/src/prover.rs:30]: remove this line

**Review by zk-steve — APPROVED** [review]

**Comment by zk-steve** [comment]
- [x] Remember to squash all commits before merging and resolve all the comments

**Review by zk-steve — APPROVED** [review]

**Pull request merged**